stream_alert/alert_merger/main.py

```diff
@@ -0,0 +1,151 @@
+            'KeyConditionExpression': Key('RuleName').eq(rule_name),
+            'Select': 'ALL_ATTRIBUTES'
+        }
+        for item in self._paginate(self.table.query, kwargs):
+            yield item
+
+    def mark_as_dispatched(self, rule_name, alert_id):
+        """Mark a specific alert as dispatched (in progress)."""
+        # Update the alerts table with the dispatch time, but only if the alert still exists.
+        # (The alert processor could have deleted the alert before the table update finishes).
+        try:
+            self.table.update_item(
+                Key={'RuleName': rule_name, 'AlertID': alert_id},
+                UpdateExpression='SET Dispatched = :now ADD Attempts :one',
+                ExpressionAttributeValues={':now': int(time.time()), ':one': 1},
+                ConditionExpression='attribute_exists(AlertID)'
+            )
+        except ClientError as error:
+            if error.response['Error']['Code'] == 'ConditionalCheckFailedException':
+                LOGGER.warn('Conditional update failed: %s', error.response['Error']['Message'])
+            else:
+                raise
+
+
+class AlertEncoder(json.JSONEncoder):
+    """Custom JSON encoder which handles sets and Decimals."""
+    def default(self, obj):  # pylint: disable=arguments-differ,method-hidden
+        if isinstance(obj, set):
+            return list(obj)
+        if isinstance(obj, Decimal):
+            return float(obj)
+        return json.JSONEncoder.default(self, obj)
+
+
+# TODO: Alert merging will be implemented here
+class AlertMerger(object):
+    """Dispatch alerts to the alert processor."""
+    def __init__(self):
+        self.alerts_db = AlertTable(os.environ['ALERTS_TABLE'])
+        self.alert_proc = os.environ['ALERT_PROCESSOR']
+        self.alert_proc_timeout = int(os.environ['ALERT_PROCESSOR_TIMEOUT_SEC'])
+        self.lambda_client = boto3.client('lambda')
+
+    def _dispatch_alert(self, alert):
+        """Dispatch all alerts which need to be sent to the rule processor."""
+        LOGGER.info('Dispatching alert %s to %s (attempt %d)',
+                    alert['AlertID'], self.alert_proc, alert.get('Attempts', 0) + 1)
+
+        self.lambda_client.invoke(
+            FunctionName=self.alert_proc,
+            InvocationType='Event',
+            Payload=json.dumps(alert, cls=AlertEncoder, separators=(',', ':')),
+            Qualifier='production'
+        )
+        self.alerts_db.mark_as_dispatched(alert['RuleName'], alert['AlertID'])
+
+    def dispatch(self):
+        """Find and dispatch all pending alerts to the alert processor."""
+        for rule_name in self.alerts_db.rule_names():
+            for alert in self.alerts_db.pending_alerts(rule_name, self.alert_proc_timeout):
+                self._dispatch_alert(alert)
+
+
+def handler(event, context):  # pylint: disable=unused-argument
+    """Entry point for the alert merger."""
+    global MERGER  # pylint: disable=global-statement
```

Comment:
I thought the `global` keyword was supposed to avoided whenever possible. We could favor class properties (not instance properties) that would keep the client active for the life of the container. I thought we were doing this elsewhere much I'm having trouble finding it.. but basically this:

```python
class Test(object):
    CLS_PROP = None

    def __init__(self):
        if not Test.CLS_PROP:
            Test.CLS_PROP = "now created"

    def get_prop(self):
        return self.CLS_PROP

for i in range(2):
    # when i == 1 the object is already cached
    print 'Class prop before init #%d: %s' % (i+1, Test.CLS_PROP)
    tester = Test()
    print 'Class prop after init #%d: %s' % (i+1, Test.CLS_PROP)
    print 'Instance prop after init #%d: %s' % (i+1, tester.get_prop())
```

I like this because it still encapsulates the client within the class that actually _uses_ it.

**EDIT**: Pardon my naivety - I was thinking this was a client wrapper but quickly realized it is not.

Reply:
No, that's a good point. After some discussion, I decided to go with your suggestion and cache the instantiation in the class itself. It makes the handler itself super clean:

```python
def handler(event, context):
    return AlertProcessor.get_instance(context.invoked_function_arn).run(event)
```